include/solana.hpp

```diff
@@ -75,6 +78,71 @@ struct Version {
  */
 void from_json(const json &j, Version &version);
 
+uint64_t trailingZeros(uint64_t n);
+uint64_t nextPowerOfTwo(uint64_t n);
+
+struct EpochSchedule {
+  uint64_t firstNormalEpoch;
+  uint64_t firstNormalSlot;
+  uint64_t leaderScheduleSlotOffset;
+  uint64_t slotsPerEpoch;
+  bool warmup;
+
+  uint64_t getEpoch(uint64_t slot) const {
+    return this->getEpochAndSlotIndex(slot)[0];
+  }
+
+  std::vector<uint64_t> getEpochAndSlotIndex(uint64_t slot) const {
+    std::vector<uint64_t> info;
+
+    if (slot < this->firstNormalSlot) {
+      const auto epoch = trailingZeros(nextPowerOfTwo(slot + MINIMUM_SLOT_PER_EPOCH + 1)) - trailingZeros(MINIMUM_SLOT_PER_EPOCH) - 1;
+      const auto epochLen = this->getSlotsInEpoch(epoch);
+      const auto slotIndex = slot - (epochLen - MINIMUM_SLOT_PER_EPOCH);
+      info.push_back(epoch);
+      info.push_back(slotIndex);
+
+      return info;
+
+    } else {
+      const auto normalSlotIndex = slot - this->firstNormalSlot;
+      const auto normalEpochIndex = floor(normalSlotIndex / this->slotsPerEpoch);
+      const auto epoch = this->firstNormalEpoch + normalEpochIndex;
+      const auto slotIndex = normalSlotIndex % this->slotsPerEpoch;
+      info.push_back(epoch);
+      info.push_back(slotIndex);
+
+      return info;
+    }
+  }
+
+  uint64_t getFirstSlotInEpoch(uint64_t epoch) const {
+    if (epoch <= this->firstNormalEpoch) {
+      return (pow(2, epoch) - 1) * MINIMUM_SLOT_PER_EPOCH;
```

Comment:
pow(2,n) can be optimized using  << on 1.

Reply:
Done